src/ethereum_test_types/block_access_list/__init__.py

```diff
@@ -0,0 +1,404 @@
+        new_instance.modifier = compose(*modifiers)
+        return new_instance
+
+    def to_fixture_bal(self, t8n_bal: "BlockAccessList") -> "BlockAccessList":
+        """
+        Convert t8n BAL to fixture BAL, optionally applying transformations.
+
+        1. First validates expectations are met (if any)
+        2. Then applies modifier if specified (for invalid tests)
+
+        Args:
+            t8n_bal: The BlockAccessList from t8n tool
+
+        Returns:
+            The potentially transformed BlockAccessList for the fixture
+
+        """
+        # Only validate if we have expectations
+        if self.account_changes:
+            self.verify_against(t8n_bal)
+
+        # Apply modifier if present (for invalid tests)
+        if self.modifier:
+            return self.modifier(t8n_bal)
+
+        return t8n_bal
+
+    def verify_against(self, actual_bal: "BlockAccessList") -> None:
+        """
+        Verify that the actual BAL from the client matches this expected BAL.
+
+        Args:
+            actual_bal: The BlockAccessList model from the client
+
+        Raises:
+            Exception: If verification fails
+
+        """
+        actual_accounts_by_addr = {acc.address: acc for acc in actual_bal.root}
+        expected_accounts_by_addr = {acc.address: acc for acc in self.account_changes}
+
+        # Check for missing accounts
+        missing_accounts = set(expected_accounts_by_addr.keys()) - set(
+            actual_accounts_by_addr.keys()
+        )
+        if missing_accounts:
+            raise Exception(
+                "Expected accounts not found in actual BAL: "
+                f"{', '.join(str(a) for a in missing_accounts)}"
+            )
+
+        # Verify each expected account
+        for address, expected_account in expected_accounts_by_addr.items():
+            actual_account = actual_accounts_by_addr[address]
+
+            try:
+                self._compare_account_changes(expected_account, actual_account)
+            except AssertionError as e:
+                raise Exception(f"Account {address}: {str(e)}") from e
+
+    def _compare_account_changes(
+        self, expected: BalAccountChange, actual: BalAccountChange
+    ) -> None:
+        """
+        Compare two BalAccountChange models with detailed error reporting.
+
+        Only validates fields that were explicitly set in the expected model,
+        using model_fields_set to determine what was intentionally specified.
+        """
+        # Only check fields that were explicitly set in the expected model
+        for field_name in expected.model_fields_set:
+            if field_name == "address":
+                continue  # Already matched by account lookup
+
+            expected_value = getattr(expected, field_name)
+            actual_value = getattr(actual, field_name)
+
+            # If we explicitly set a field to None, verify it's None/empty
+            if expected_value is None:
```

Comment:
I think `BalAccountChange` as is right now does not allow `None` in any of its fields, so it could be a bit annoying because type checker is going to complain a lot, but I think the tester should anyway use `[]` to explicitly state that the fields needs to be an empty list instead of `None`.

Reply:
> I think the tester should anyway use [] to explicitly state that the fields needs to be an empty list instead of None.

I think this is the right approach. We can relax it to accept both `None` or `[]` but I'm not sure that's great either. I went with raising with a good message to use empty lists on `None`. Thanks for catching this, lot's of changes to keep track. This was a `None` check at one point 😅. I will write some basic unit tests for this.